File: aoc-py/solutions/day6.py
"""
Day 6: Wait For It

https://adventofcode.com/2023/day/6
"""
__all__ = ('Day6',)

from math import prod, sqrt, ceil

from ..solution import Solution

class Day6(Solution):
    NAME = 'Wait For It'

    def _get_num_beats_bf(self, time: int, to_beat: int) -> int:
        """Brute forces the amount of ways in a race of `time` milliseconds
        to beat the record distance of `to_beat`

        Loops throught all possible values that one could hold the button for: `hold_time`
        and checks if the distance travelled is greater than the distance to beat
            = (speed * travel_time)
            where speed = hold_time
                (the amount of time the button is held directly equates to the speed)
            and travel_time = (time - hold_time)
                (The amount of time we have left to move; since we can only move after we stop holding the button)
            therefore our conditional is:
                hold_time * (time - hold_time) > to_beat
                and everytime it holds true, it increases the amount of combinations that beat `to_beat`
        """
        return sum(
            1 for hold_time in range(1, time)
            if hold_time * (time - hold_time) > to_beat
        )

    def _get_num_beats(self, time: int, to_beat: int) -> int:
        """Using Quadratics
    
        let d(x) be the distance travelled
            where x = the time the button is held
        d(x) = x * (t - x)
            where t = the amount of time the race lasts, t > 0
        d(x) = -x^2 + tx

        we need to calculate where d(x) > r
            where r = the record distance that we need to beat
        -x^2 + tx > r
        -x^2 + tx - r > 0

        let r1, r2 be the 2 roots of the equation respectively by the quadratic formula
            in ax^2 + bx + c:
                a = -1
                b = t
                c = -r
            therefore:
                r = -t +- √[t^2 - 4(-1)(-r)] / [2 * (-1)]
            and
                r1 = t - √(t^2 + 4r) / 2
                r2 = t + √(t^2 + 4r) / 2
        therefore:
            r1 < x < r2 is our solution to the inequality
            and total amount of winning possibilities would simply be ⌊r2⌋ - ⌈r1⌉ + 1
            which are the number of integers that satisfy the inequality
        """
        discriminant = time ** 2 - 4 * to_beat
        if discriminant < 0:
            return 0
        root1 = (time - sqrt(discriminant)) / 2
        root2 = (time + sqrt(discriminant)) / 2
        return int(root2) - ceil(root1) + 1

    def part_one(self, inp: str) -> int:
        time, distance = inp.splitlines()
        _, time = time.split(':', maxsplit=1)
        _, distance = distance.split(':', maxsplit=1)

        times = map(int, time.split())
        distances = map(int, distance.split())

        return prod(
            self._get_num_beats(time, to_beat)
            for time, to_beat in zip(times, distances)
        )

    def part_two(self, inp: str) -> int:
        time, distance = inp.replace(' ', '').splitlines()
        _, time = time.split(':', maxsplit=1)
        _, distance = distance.split(':', maxsplit=1)

        return self._get_num_beats(
            int(time),
            int(distance),
        )

    def part_one_bf(self, inp: str) -> int:
        """Brute force solution for part 1"""

        times, distances = inp.splitlines()
        _, times = times.split(':', maxsplit=1)
        _, distances = distances.split(':', maxsplit=1)

        times = map(int, times.split())
        distances = map(int, distances.split())

        return prod(
            self._get_num_beats_bf(time, to_beat)
            for time, to_beat in zip(times, distances)
        )

    def part_two_bf(self, inp: str) -> int:
        """Brute force solution for part 2"""

        time, distance = inp.replace(' ', '').splitlines()
        _, time = time.split(':', maxsplit=1)
        _, distance = distance.split(':', maxsplit=1)

        return self._get_num_beats_bf(
            int(time),
            int(distance),
        )

    def run(self, inp: str) -> None:
        print('Part 1:', p1 := self.part_one(inp))
        print('Part 2:', p2 := self.part_two(inp))

        assert p1 == 1731600
        assert p2 == 40087680

File: src/bin/day6.rs
//! Day 6: Wait For It
//!
//! <https://adventofcode.com/2023/day/6>
use std::fmt::Display;
use aoc_2023::Solution;

pub struct Day6;

impl Day6 {
    /// Brute forces getting the amount of cases where we beat the target distance
    /// by looping through all the possible values for the time the button is held,
    /// calculating the distance based on that and checking if it beats the target distance that we need to beat
    #[inline]
    fn get_num_beats_bf((time, to_beat): (usize, usize)) -> usize {
        (1..time)
            .filter(|hold_time| hold_time * (time - hold_time) > to_beat)
            .count()
    }

    /// Mathematical method of getting the amount of cases where we beat the target distance
    ///
    /// The function d(t) which models the distance travelled based on the time the button is held is a quadratic
    /// we simply need to solve the inequality were d(t) > [the time we need to beat]
    /// then we simply need to calculate the amount of integers in between the range of the inequality
    /// AKA the 2 roots of the function d(t) - [the time we need to beat]
    /// that number would be ⌊greater root⌋ - ⌈smaller root⌉ + 1
    #[allow(
        clippy::cast_sign_loss,
        clippy::cast_precision_loss,
        clippy::cast_possible_truncation,
    )]
    fn get_num_beats((time, to_beat): (usize, usize)) -> usize {
        let time = time as f64;
        let to_beat = to_beat as f64;
        let discriminant = 4.0_f64
            .mul_add(-to_beat, time.powi(2));

        if discriminant < 0.0 { 0 }
        else {
            let root1 = (time - discriminant.sqrt()) / 2.0;
            let root2 = (time + discriminant.sqrt()) / 2.0;
            (root2.trunc() - root1.ceil() + 1.0) as usize
        }
    }

    /// Helper function for Part 1 that executes the brute force and non brute force methods
    fn part_one_helper<T, F>(inp: T, map_func: F) -> usize
    where
        T: Display,
        F: Fn((usize, usize)) -> usize,
    {
        inp
            .to_string()
            .replace('\r', "")
            .split_once('\n')
            .map(|(time, distance)| (
                time.split_once(':')
                    .unwrap()
                    .1
                    .split_whitespace()
                    .filter_map(|n| n.parse::<usize>().ok())
                    .zip(
                        distance.split_once(':')
                            .unwrap()
                            .1
                            .split_whitespace()
                            .filter_map(|n| n.parse::<usize>().ok())
                    )
                    .map(map_func)
                    .product()
                )
            )
            .unwrap()
    }

    /// Helper function for Part 2 that executes the brute force and non brute force methods
    fn part_two_helper<T, F>(inp: T, map_func: F) -> usize
    where
        T: Display,
        F: Fn((usize, usize)) -> usize,
    {
        inp
            .to_string()
            .replace(|c| c == '\r' || c == ' ', "")
            .split_once('\n')
            .map(|(time, distance)|
                map_func((
                    time.split_once(':')
                        .unwrap()
                        .1
                        .parse::<usize>()
                        .unwrap(),
                    distance.split_once(':')
                        .unwrap()
                        .1
                        .parse::<usize>()
                        .unwrap(),
                ))
            )
            .unwrap()
    }

    /// Non brute force part 1
    pub fn part_one<T: Display>(&self, inp: T) -> usize {
        Self::part_one_helper(inp, Self::get_num_beats)
    }

    /// Brute force part 1
    pub fn part_one_bf<T: Display>(&self, inp: T) -> usize {
        Self::part_one_helper(inp, Self::get_num_beats_bf)
    }

    /// Non brute force part 2
    pub fn part_two<T: Display>(&self, inp: T) -> usize {
        Self::part_two_helper(inp, Self::get_num_beats)
    }

    /// Brute force part 2
    pub fn part_two_bf<T: Display>(&self, inp: T) -> usize {
        Self::part_two_helper(inp, Self::get_num_beats_bf)
    }
}

impl Solution for Day6 {
    const NAME: &'static str = "Wait For It";

    fn run(&self, inp: String) {
        let p1 = self.part_one(&inp);
        let p2 = self.part_two(&inp);

        assert_eq!(self.part_one_bf(&inp), p1);
        assert_eq!(self.part_two_bf(&inp), p2);

        println!("Part 1: {p1}");
        println!("Part 2: {p2}");

        assert_eq!(p1, 1_731_600);
        assert_eq!(p2, 40_087_680);
    }
}

fn main() {
    aoc_2023::run_day(6, &Day6);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() { main(); }
}